src/Daemon.hs

```diff
@@ -67,13 +70,28 @@ msgDaemon
     -> String
     -> (a -> IO ())
     -> (SomeException -> IO ())
+    -> Bool
     -> IO (MsgDaemon a)
-msgDaemon logger name computation handleException = do
+msgDaemon logger name computation handleException threadSafe = do
     chan <- newTChanIO
+    alreadyRunningRef :: MVar (Maybe ThreadId) <- newMVar Nothing
 
     let sendMsg = atomically . writeTChan chan
-        loop = forkIO . forever $ run `catch` handle
+        loop = do
+            alreadyRunning <- takeMVar alreadyRunningRef
+            case (threadSafe, alreadyRunning) of
+                (False, Just tid) -> do
```

Comment:
If a daemon is thread safe, then we can behave as before.  If it is *not*, we make it so here.  See call in `logDaemon` below.